libgadget/exchange.c

```diff
@@ -169,11 +169,19 @@ static int domain_exchange_once(int (*layoutfunc)(int p), ExchangePlan * plan, i
 
     int bad_exh=0;
 
-    /*Check whether the domain exchange will succeed. If not, bail*/
+    /*Check whether the domain exchange will succeed. If not, check for garbage and then check again.*/
     if(PartManager->NumPart + plan->toGetSum.base - plan->toGoSum.base > PartManager->MaxPart){
-        message(1,"Too many particles for exchange: NumPart=%d count_get = %d count_togo=%d MaxPart=%d\n",
+        /* How much garbage have we got? Keep this serial so there is no memory allocation.
```

Comment:
Is the idea the garbages will be always be freed so we actually have more free slots?

Reply:
That's right - we have to do a gc anyway to get rid of the particles that are leaving this processor, and the garbage gets swept up in that. But hang on, I can do this a better way.